include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,585 @@
+ * some part of the domain (e.g. around a crack tip) and use standard FE (e.g. FE_Q)
+ * elsewhere. In order to make the resulting FE space $C^0$ continuous in the
+ * case of continuous enrichmenet functions, all enriched DoFs on the interface
+ * between enriched and non-enriched elements are constrained to be zero.
+ * To get those constraints working automatically, the user needs to wrap a FE
+ * to be used in part of the domain without enrichment:
+ * @code
+ *   FE_Enriched<dim> fe_non_enriched(FE_Q<dim>(1));
+ * @endcode
+ * This constructor is equivalent to calling
+ * @code
+ *   FE_Enriched<dim> fe_non_enriched(FE_Q<dim>(1),
+ *                                    FE_Nothing<dim>(1,true),
+ *                                    NULL);
+ * @endcode
+ * and will result in the correct constraints for enriched DoFs attributed to
+ * support points on the interface between the two regions.
+ *
+ * <h3>Implementation</h3>
+ *
+ * The implementation of the class is based on FESystem which is aggregated as
+ * a private member. The simplest constructor <code> FE_Enriched<dim> fe(FE_Q<dim>(2), FE_Q<dim>(1),function)</code>
+ * will internally initialize FESystem as
+ *
+ * @code
+ * FESystem<dim> fe_system(FE_Q<dim>(2),1,
+ *                         FE_Q<dim>(1),1);
+ * @endcode
+ *
+ * Note that it would not be wise to have this class derived
+ * from FESystem as the latter concatenates the given elements into different
+ * components of a vector element, whereas the current class combines the given
+ * elements into the same components. For instance, if two scalar elements are
+ * given, the resulting element will be scalar rather than have two components
+ * when doing the same with an FESystem.
+ *
+ * The ordering of the shape function, @p interface_constrains, the @p prolongation (embedding)
+ * and the @p restriction matrices are taken from the FESystem class.
+ *
+ * <h3>References</h3>
+ *
+ * When using this class, please cite
+ * @code{.bib}
+ @Article{Davydov2016,
+    author =    {Denis Davydov and Tymofiy Gerasimov and Jean-Paul Pelteret and Paul Steinmann},
+    title =     {On the h-adaptive PUM and hp-adaptive FEM approaches applied to PDEs in quantum mechanics.},
+    journal =   {Journal of Computational Physics. Submitted},
+    year =     {2016},
+ }
+ * @endcode
+ * The PUM was introduced in
+ * @code{.bib}
+ @Article{Melenk1996,
+  Title                    = {The partition of unity finite element method: Basic theory and applications },
+  Author                   = {J.M. Melenk and I. Babu\v{s}ka},
+  Journal                  = {Computer Methods in Applied Mechanics and Engineering},
+  Year                     = {1996},
+  Number                   = {1--4},
+  Pages                    = {289 -- 314},
+  Volume                   = {139},
+}
+@Article{Babuska1997,
+  Title                    = {The partition of unity method},
+  Author                   = {Babu\v{s}ka, I. and Melenk, J. M.},
```

Comment:
Just a minor nitpick, all three of these references differ in the way they abbreviate (or not) first names, and where they appear relative to last names.